docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,308 @@
+---
+title: Manifest Hydrator
+authors:
+  - "@crenshaw-dev"
+  - "@zachaller"
+sponsors:
+  - TBD        # List all interested parties here.
+reviewers:
+  - TBD
+approvers:
+  - TBD
+
+creation-date: 2024-03-26
+last-updated: 2024-03-26
+---
+
+# Manifest Hydrator
+
+This proposal describes a feature to make manifest hydration (i.e. the "rendered manifest pattern") a first-class feature of Argo CD.
+
+## Open Questions 
+
+* The `sourceHydrator` field is mutually exclusive with the `source` and the `sources` field. Should we throw an error if they're both configured, or should we just pick one and ignore the others?
```

Comment:
Can you define CEL for me? But yep, I generally feel like an error message is better.

Reply:
https://kubernetes.io/docs/reference/using-api/cel/

Reply:
Using CEL in the CRD spec should allow the **k8s API server** to reject objects when both fields are set, as an alternative to accepting and persisting the objects and only then throwing the error in ArgoCD controller log.

Reply:
Gotcha. I think CEL would be out of scope for this proposal since it would presumably be a fairly significant effort to get Argo CD to OpenAPI v3. But it's definitely much nicer than a runtime error!